jwe.go

```diff
@@ -0,0 +1,80 @@
+package jwe
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"strings"
+)
+
+// NewJWE creates a new JWE token.
+// The plaintext will be encrypted with the method using a cek(Content Encryption Key).
+// The cek will be encrypted with the alg using the key.
+func NewJWE(alg KeyAlgorithm, key interface{}, method EncryptionType, plaintext []byte) (*jwe, error) {
+	jwe := &jwe{}
+
+	jwe.protected.Enc = method
+	chipher, err := getCipher(method)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate a random Content Encryption Key (CEK).
+	cek, err := generateKey(chipher.keySize)
+	if err != nil {
+		return nil, err
+	}
+
+	// Encrypt the CEK with the recipient's public key to produce the JWE Encrypted Key.
+	jwe.protected.Alg = alg
+	encrypter, err := createEncrypter(key)
+	if err != nil {
+		return nil, err
+	}
+	jwe.recipientKey, err = encrypter.Encrypt(cek, alg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Serialize Authenticated Data
+	rawProtected, err := json.Marshal(jwe.protected)
+	if err != nil {
+		return nil, err
+	}
+	rawProtectedBase64 := base64.RawURLEncoding.EncodeToString(rawProtected)
+
+	// Perform authenticated encryption on the plaintext
+	jwe.iv, jwe.ciphertext, jwe.tag, err = chipher.encrypt(cek, []byte(rawProtectedBase64), plaintext)
+	if err != nil {
+		return nil, err
+	}
+
+	return jwe, nil
+}
+
+type jwe struct {
```

Comment:
even though this is unexported, this is the most valuable struct and probably deserves a quick doc comment.

Reply:
Added